Backend work for public collections: thumbnails, url list, upload pages, and so on

tw4l opened the pull request:
Fixes #2182 

This rather large PR adds the rest of what should be needed for public collections work in the frontend.

## New API endpoints

- `/public/orgs/{org_slug}/collections/{coll_id}`: Public GET endpoint for public collection (returns 404 `collection_not_found` if org with slug doesn't exist or isn't public, collection id doesn't exist or isn't public)
- - `/public/orgs/{org_slug}/collections/{coll_id}/download`: Streaming download endpoint for public collection (returns 404 `collection_not_found` if org with slug doesn't exist or isn't public, collection id doesn't exist or isn't public; 403 if `Collection.allowPublicDownload` is false)
- `/orgs/{oid}/collections/{coll_id}/urls`: Paginated GET list of urls in collection sorted descending by snapshot count, with information about each individual snapshot (including page_id, needed for below), for use in selecting a home url/snapshot for collections in the frontend
- `/orgs/{oid}/collections/{coll_id}/home-url`: POST endpoint to set collection home url/start snapshot by page_id
- `/orgs/{oid}/collections/{coll_id}/thumbnail`: PUT endpoint to upload collection thumbnail as a stream. Will replace existing thumbnail if called on a collection that already has one.
- `/orgs/{oid}/collections/{coll_id}/thumbnail`: DELETE endpoint to remove collection thumbnail

## Changes to existing API endpoints

- `/public-collections/{org_slug}`: Public collection list now supports pagination, only returns public fields for collections, and has had the path modified to `/public/orgs/{org_slug}/collections` following offline feedback
- Several `pages` endpoints that previously only supported `/crawls/` in their path, e.g. `/orgs/{oid}/crawls/all/pages/reAdd`, now support `/uploads/` and `/all-crawls/` namespaces as well. This is necessitated by adding pages for uploads to the database (see below). For `/orgs/{oid}/namespace/all/pages/reAdd`, `crawls` or `uploads` will serve as a filter to only affect crawls of that given type. Other endpoints are more liberal at this point, and will perform the same action regardless of the namespace used in the route (we'll likely want to change this in a follow-up to be more consistent, but I didn't want to expand the scope of this PR even larger than it already has been).
- `/orgs/{oid}/namespace/all/pages/reAdd` now kicks off a background job rather than doing all of the computation in an asyncio task in the backend container. The background job additionally updates collection date ranges, page/size counts, and tags for each collection in the org after pages have been (re)added.
- The collection PATCH endpoint now allows the user to set a default thumbnail that is served by the frontend via the `defaultThumbnailName` field, as well as the new `allowPublicDownload` field to determine whether the collection is downloadable when/if it is made public.

## Other big changes

- New uploads will now have their pages read into the database! Collection page counts now also include uploads!
- A migration was added to start a background job for each org that will add the pages for previously-uploaded WACZ files to the database and update collections accordingly!
- Adds a new `ImageFile` subclass of `BaseFile` for thumbnails that we can use for other user-uploaded image files moving forward, with separate output models for authenticated and public endpoints

## Other less big changes

- Collections now have a `dateEarliest` and `dateLatest` that we can generate a range from, based on the earliest and latest snapshot in the collection
- Public collections now have their own model that restricts the fields we share to unauthenticated users
- Collections now have a `caption` field that can be used to set a short caption to use for public collection views (presumably with more visibility than full description)
- The process of adding pages to the database from WACZ files is more flexible to account for older WACZ files whose page lists may deviate from the current norm (e.g. non-UUID page ids)

## Testing

Backend tests have been added or modified for all API endpoint changes.

To test the migration:
- Spin up `main` locally
- Upload some WACZ files in one or more orgs and add them to some collections
- Re-deploy from this branch
- Verify that migration appeared to work successfully in backend logs
- Verify that background job container (in default namespace) logs look good
- Verify that pages have been added for uploads and that collections have been updated accordingly (look at data in Network tab for collection replay.json endpoint on the collection page)

tw4l committed:
Add list endpoint to get sorted list of URLs in collection

Response is sorted desc by page count match and includes an array
containing page_id, ts, and status for each snapshot with that URL.

tw4l committed:
Add endpoint to set or update collection home url

tw4l committed:
Fixups

tw4l committed:
Use updated response for /home-urls endpoint

tw4l committed:
Add tests

tw4l committed:
WIP: Add collection thumbnail upload

Also reorganizes some utility classes and collection get methods.

Caveats:
- Generates a new presigned url for the thumbnail image each time
CollOut is created (may want to do something similar to crawl files
and track expiration date)

tw4l committed:
Add collection thumbnail upload test

tw4l committed:
WIP: Only share subset of coll and image data for public endpoint

TODO:
- Add data ranges to collections (maybe precompute on base model)
- Add caption for public collections?

tw4l committed:
Fix test that was checking wrong collection

tw4l committed:
Change path for thumbnail endpoint

tw4l committed:
Add endpoint to delete collection thumbnail

Also renames StorageOps.delete_crawl_file_object to simply
delete_file_object, since we use it across all types of files we
store in s3.

tw4l committed:
Add caption field to collections

tw4l committed:
Calculate and store earliest and latest dates in collection

- Based on page timestamps
- Not all archived items have pages written to the db yet, so this
won't count pages in older crawls and in uploads yet
- Might want to move to asyncio task or background job

tw4l committed:
Add pages to database for uploads

Note:
- Page endpoints have crawls in path, not uploads or all-crawls,
but will now work with uploads
- To add pages from older uploads, the pages endpoints to re-add
pages for a particular crawl or all crawls will now work for
uploads as well

tw4l committed:
WIP: Add all-crawls/ and uploads/ versions of GET pages endpoints

- QA-related endpoints are still restricted to crawls only for
now
- Do we want to enforce the collection types or are these multiple
paths per endpoint method sufficient? tbd

tw4l committed:
Add comment to get CI to run

tw4l committed:
Fix test

tw4l committed:
Add fallbacks if pages have no id or a non-UUID id

tw4l committed:
Move pages test later to give it more time

tw4l committed:
Delete upload pages when deleted or replaced

tw4l committed:
Remove asserts for upload pages for optional fields

tw4l committed:
Remove outdated comment

tw4l committed:
Filter (re-)add all pages endpoint by crawl type in path

If a user POSTS to /crawls/all/ or /uploads/all/, filter the
crawl objects in the database by that type before adding pages.
This enables us to easily add pages to the database for all
uploads in an organization, for example.

GET endpoints and endpoints that specify just a single crawl have
been kept more permissive as long as the crawl_id is found.

tw4l committed:
Move re-adding all pages in org to background job

tw4l committed:
Add migration to add upload files for all orgs in bg jobs

tw4l committed:
Add API endpoint to GET single public collection

tw4l committed:
Recalculate collection dates after adding upload pages

tw4l committed:
Rename type_filter to crawl_type

tw4l committed:
Recalculate collection stats after adding upload pages

tw4l committed:
Reduce per-page print logging

tw4l committed:
Include upload pages in collection pageCount

tw4l committed:
Return 404 for GET public collection if org isn't public

SuaYoo commented on backend/btrixcloud/models.py:
```diff
@@ -1063,34 +1230,86 @@ class Collection(BaseMongoModel):
+class PublicCollOut(BaseMongoModel):
+    """Collection output model with annotations."""
+
+    name: str
```
We'll need `oid` in the public response to construct the replay link

tw4l replied:
Added!

tw4l committed:
Add oid to PublicCollOut model

tw4l committed:
Modify public collections endpoint paths

Use public namespace, with form:
- /public/orgs/slug/collections
- /public/orgs/slug/collections/collection_id

tw4l committed:
Fix endpoint path in test

SuaYoo commented on backend/btrixcloud/models.py:
```diff
@@ -1072,16 +1258,59 @@ class CollIn(BaseModel):
+    homeUrl: Optional[AnyHttpUrl] = None
+    homeUrlTs: Optional[datetime] = None
+
     resources: List[CrawlFileOut] = []
```
I missed this, but [from the mockups](https://private-user-images.githubusercontent.com/4672952/390177135-415d1b27-9b32-4d22-92a4-646b40bbd6c9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMyODQzNjQsIm5iZiI6MTczMzI4NDA2NCwicGF0aCI6Ii80NjcyOTUyLzM5MDE3NzEzNS00MTVkMWIyNy05YjMyLTRkMjItOTJhNC02NDZiNDBiYmQ2YzkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMDRUMDM0NzQ0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDNjNGE3Y2JmNzRhOWY2MTZkYTY4NWU0ZjA1ZGNmZjI5NzQxYzM0N2VkODU3ZDg3ZDZmMDdhYWUxOTkyMGYxMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.kyngjs_kr6sSojLrhz6H-PEKKPtRHh8t9Mj8bzROWY0) we'll also need to return `totalSize`, `pageCount`, and `crawlCount`.

tw4l committed:
Only fetch org once, not per-crawl

tw4l committed:
Add counts and size to PublicCollOut

tw4l committed:
Enforce max thumbnail file size of 2MB

tw4l committed:
Use proper thumbnail as test data

tw4l commented:
@SuaYoo I added a max size validation for thumbnails of 2 MB (following Youtube's example). It'll be easy to raise it or set different limits for other types of user-uploaded image files in the future. I'm wondering if we also want to restrict file types? I'm calculating a mime type for the image files based on the filename they're uploaded with, we could use that or try to detect via magic numbers

tw4l committed:
Allow unlisted collections in single public collection GET

tw4l committed:
Add access field to PublicCollOut

tw4l committed:
Don't expect thumbnail for unlisted collection

tw4l committed:
Fix thumbnail-related 500 error with private collections list

tw4l committed:
Add defaultThumbnailName field to collections

This allows the frontend to specify which default thumbnail to use
without needing to re-upload the same file to the backend and
stream if from s3 each time it's used.

The field can be set on collection creation and is is updatable
via the collection PATCH endpoint.

tw4l committed:
Add fallback values for access in coll out models

tw4l committed:
Add ability to unset home url via /home-url endpoint

tw4l committed:
Add public collection download endpoint

tw4l committed:
Implement and enforce Collection.allowPublicDownload

tw4l committed:
Fix linting

tw4l committed:
Make sure coll out models return allowPublicDownload as bool

SuaYoo reviewed (approved):
Based on the availability of API endpoints, this looks like it covers everything we need on the backend so far! I'm also able to get it running locally.

tw4l committed:
Make migration idempotent - don't readd existing upload pages

tw4l committed:
Reformat migration

tw4l committed:
Fix typing

SuaYoo commented:
Documenting some feedback from Discord:
- Ran into an issue when using the `/urls` endpoint with prefix search. `urlPrefix` doesn't seem to accept values after `?`, for example `[frontend/src/features/collections/select-collection-start-page.ts](https://en.wikipedia.org/w/index.php?title=Special:CreateAccount&returnto=Main+Page)`. Passing in an encoded value to `urlPrefix` doesn't return any results.
- Unlisted collections should still return in the public GET endpoint, regardless of whether the org itself is public.

tw4l committed:
Allow getting and downloading public collections if org profile disabled

tw4l committed:
URL decode urlPrefix in /urls endpoint

tw4l committed:
Do URL decoding inside list urls method

tw4l committed:
Escape special characters in url_prefix regex

tw4l commented:
> Documenting some feedback from Discord:
> 
>     * Ran into an issue when using the `/urls` endpoint with prefix search. `urlPrefix` doesn't seem to accept values after `?`, for example `[frontend/src/features/collections/select-collection-start-page.ts](https://en.wikipedia.org/w/index.php?title=Special:CreateAccount&returnto=Main+Page)`. Passing in an encoded value to `urlPrefix` doesn't return any results.
> 
>     * Unlisted collections should still return in the public GET endpoint, regardless of whether the org itself is public.

@SuaYoo these are both now fixed, pushed to dev, and tested

SuaYoo reviewed (approved):
Tested newest changes locally, working well!

SuaYoo commented:
@tw4l mind if I merge this into the feature branch?

tw4l commented:
> @tw4l mind if I merge this into the feature branch?

done! :)
